Compatibility with PHP 8.1

srjlewis opened the pull request:
Add typing and compatibility with PHP 8.1

I have also pulled in #191

Commit:
allow psr/log v2 (#191)

* allow psr/log v2

widens the restriction for which logger interface to use.
similar to how symfony handles this: https://github.com/symfony/console/blob/5.3/composer.json#L34

srjlewis committed:
php 8.1 support

srjlewis committed:
Merge remote-tracking branch 'origin/3.x-actions' into 5.x-php8.1

srjlewis committed:
php 8.1 support

srjlewis committed:
php 8.1 support

srjlewis committed:
php 8.1 support

srjlewis committed:
Merge remote-tracking branch 'origin/3.x' into 5.x-php8.1

# Conflicts:
#	composer.json

srjlewis committed:
Remove JetBrains\PhpStorm\ArrayShape usage

srjlewis committed:
Fixing regression in composer.json with phpunit

kenjis commented:
@koriym Can you review?

koriym commented on src/ConnectionLocator.php:
```diff
@@ -42,21 +42,21 @@ class ConnectionLocator implements ConnectionLocatorInterface
      * Constructor.
      *
-     * @param callable $default A callable to create a default connection.
+     * @param callable|null $default A callable to create a default connection.
```
Isn't "?callable" good enough?

srjlewis replied:
nullable notation `?<type>` is not recognized within docblock by static analyzers such as phpStan.

phpStan Docs https://phpstan.org/writing-php-code/phpdoc-types
phpdoc guide https://docs.phpdoc.org/3.0/guide/guides/types.html

So I am using union type notation `<type>|null` as shown in the above docs.

koriym replied:
Got it. Thanks for the clarification.

srjlewis committed:
correctly set minimum php version in composer

kenjis commented on composer.json:
```diff
@@ -22,8 +22,9 @@
     "require": {
-        "php": ">=5.6.0",
-        "psr/log": "^1.0"
+        "php": ">=8.0",
```
Aura 5.x requires PHP 8.1 or later.

```suggestion
        "php": ">=8.1",
```

srjlewis replied:
It does work fine on php 8.0, but I am fine with it being set to 8.1, ~~did you mean to remove the psr/log requirement?~~ sorry miss read the change

kenjis replied:
Aura 5.x does not support PHP 8.0.
PHP 8.0 is supported by Aura 4.x or 3.x.

I did not mean remove the psr/log.

srjlewis committed:
Update composer.json to php 8.1

Co-authored-by: kenjis <kenji.uui@gmail.com>

kenjis commented on .github/workflows/tests.yml:
```diff
@@ -5,7 +5,7 @@ jobs:
     strategy:
       matrix:
-        php-versions: ['5.6', '7.0', '7.1', '7.2', '7.3', '7.4', '8.0']
+        php-versions: ['8.0', '8.1']
```
```suggestion
        php-versions: ['8.1']
```

srjlewis replied:
I was just about do that change.

srjlewis committed:
Update .github/workflows/tests.yml to only test on with php 8.1

Co-authored-by: kenjis <kenji.uui@gmail.com>

srjlewis committed:
Update test for a change in behavior 

By default PDO now uses native types for int/floats instead of everything being strings.

srjlewis committed:
correct docblock to match type

srjlewis committed:
test null default connection within ConnectionLocator and throw ConnectionNotFound from getDefault() on a null default

srjlewis reviewed (commented):
Add a extra test, all looks good

srjlewis committed:
Update src/PdoInterface.php

Co-authored-by: kenjis <kenji.uui@gmail.com>

srjlewis committed:
Update src/PdoInterface.php

Co-authored-by: kenjis <kenji.uui@gmail.com>

srjlewis committed:
Update src/PdoInterface.php

Co-authored-by: kenjis <kenji.uui@gmail.com>

srjlewis committed:
Update src/PdoInterface.php

Co-authored-by: kenjis <kenji.uui@gmail.com>

kenjis commented on src/PdoInterface.php:
```diff
@@ -185,5 +186,5 @@ public function setAttribute($attribute, $value);
      *
      */
-    public static function getAvailableDrivers();
+    //public static function getAvailableDrivers(): array;
```
Why commented out?

srjlewis replied:
I think I commented it out as static analysis showed it was not used and no tests was falling. I will be adding it back with a corresponding test.

srjlewis committed:
Update src/PdoInterface.php

Co-authored-by: kenjis <kenji.uui@gmail.com>

srjlewis committed:
Update src/PdoInterface.php

Co-authored-by: kenjis <kenji.uui@gmail.com>

srjlewis committed:
Update src/AbstractExtendedPdo.php

Co-authored-by: kenjis <kenji.uui@gmail.com>

srjlewis committed:
Apply suggestions from code review

Co-authored-by: kenjis <kenji.uui@gmail.com>

kenjis commented on src/PdoInterface.php:
```diff
@@ -70,18 +70,18 @@ public function errorInfo();
      *
      */
-    public function exec($statement);
+    public function exec(string $statement): int;
```
```suggestion
    public function exec(string $statement): int|false;
```

kenjis commented on src/PdoInterface.php:
```diff
@@ -70,18 +70,18 @@ public function errorInfo();
      *
      */
-    public function getAttribute($attribute);
+    public function getAttribute(int $attribute): mixed;
```
https://www.php.net/manual/en/pdo.getattribute.php
```suggestion
    public function getAttribute(int $attribute): bool|int|string|array|null;
```

kenjis commented on src/PdoInterface.php:
```diff
@@ -147,12 +148,12 @@ public function query($statement, ...$fetch);
      *
      */
-    public function quote($value, $parameter_type = PDO::PARAM_STR);
+    public function quote(mixed $value, int $parameter_type = PDO::PARAM_STR): string|false;
```
```suggestion
    public function quote(string|int|array|float|null $value, int $type = PDO::PARAM_STR): string|false;
```

kenjis commented on src/PdoInterface.php:
```diff
@@ -147,12 +148,12 @@ public function query($statement, ...$fetch);
      *
      * @param int $parameter_type A data type hint for the database driver.
```
```suggestion
     * @param int $type A data type hint for the database driver.
```

kenjis commented on src/PdoInterface.php:
```diff
@@ -117,27 +117,28 @@ public function lastInsertId($name = null);
      *
      */
-    public function query($statement, ...$fetch);
+    public function query($statement, $fetchMode = null, ...$fetch_mode_args): PDOStatement|false;
```
```suggestion
    public function query(string $query, ?int $fetchMode = null, ...$fetch_mode_args): PDOStatement|false;
```

kenjis commented on src/PdoInterface.php:
```diff
@@ -117,27 +117,28 @@ public function lastInsertId($name = null);
      *
      * Queries the database and returns a PDOStatement.
      *
      * @param string $statement The SQL statement to prepare and execute.
```
```suggestion
     * @param string $query The SQL statement to prepare and execute.
```

kenjis commented on src/PdoInterface.php:
```diff
@@ -117,27 +117,28 @@ public function lastInsertId($name = null);
-     *
      */
-    public function prepare($statement, $options = null);
+    public function prepare(string $statement, array $options = []): PDOStatement|false;
```
```suggestion
    public function prepare(string $query, array $options = []): PDOStatement|false;
```

kenjis commented on src/AbstractExtendedPdo.php:
```diff
@@ -574,7 +572,7 @@ public function perform($statement, array $values = [])
      *
      */
-    public function prepare($statement, $options = [])
+    public function prepare(string $statement, array $options = []): PDOStatement
```
```suggestion
    public function prepare(string $query, array $options = []): PDOStatement
```

kenjis commented on src/AbstractExtendedPdo.php:
```diff
@@ -574,7 +572,7 @@ public function perform($statement, array $values = [])
+    public function prepare(string $statement, array $options = []): PDOStatement
     {
         $this->connect();
         $sth = $this->pdo->prepare($statement, $options);
```
```suggestion
        $sth = $this->pdo->prepare($query, $options);
```

kenjis commented on src/AbstractExtendedPdo.php:
```diff
@@ -635,18 +634,20 @@ public function prepareWithValues($statement, array $values = [])
      *
      */
-    public function query($statement, ...$fetch)
+    public function query($statement, $fetchMode = null, mixed ...$fetch_mode_args): PDOStatement
```
```suggestion
    public function query(string $query, ?int $fetchMode = null, mixed ...$fetch_mode_args): PDOStatement
```

kenjis commented on src/AbstractExtendedPdo.php:
```diff
@@ -635,18 +634,20 @@ public function prepareWithValues($statement, array $values = [])
         $this->connect();
         $this->profiler->start(__FUNCTION__);
-        $sth = $this->pdo->query($statement, ...$fetch);
+        $sth = $this->pdo->query($statement, $fetchMode, ...$fetch_mode_args);
```
```suggestion
        $sth = $this->pdo->query($query, $fetchMode, ...$fetch_mode_args);
```

kenjis commented on src/AbstractExtendedPdo.php:
```diff
@@ -667,10 +668,12 @@ public function query($statement, ...$fetch)
      *
      */
-    public function quote($value, $type = self::PARAM_STR)
+    public function quote(mixed $value, int $type = self::PARAM_STR): string
```
```suggestion
    public function quote(string|int|array|float|null $value, int $type = self::PARAM_STR): string
```

kenjis commented on src/AbstractExtendedPdo.php:
```diff
@@ -987,7 +990,7 @@ protected function setQuoteName($driver)
      * @return mixed
      */
-    public function getAttribute($attribute)
+    public function getAttribute(int $attribute): mixed
```
```suggestion
    public function getAttribute(int $attribute): bool|int|string|array|null
```

kenjis commented on src/AbstractExtendedPdo.php:
```diff
@@ -1001,7 +1004,7 @@ public function getAttribute($attribute)
      * @return bool
      */
-    public function setAttribute($attribute, $value)
+    public function setAttribute(int $attribute, mixed$value): bool
```
```suggestion
    public function setAttribute(int $attribute, mixed $value): bool
```

kenjis commented:
Could you update the config file of PHPUnit?

```
PHPUnit 9.5.10 by Sebastian Bergmann and contributors.

  Warning - The configuration file did not pass validation!
  The following problems have been detected:

  Line 3:
  - Element 'testsuite': The attribute 'name' is required but missing.

  Line 7:
  - Element 'filter': This element is not expected.

  Test results may not be as expected.
```

srjlewis committed:
update phpunit config

srjlewis committed:
re-add PdoInterface::getAvailableDrivers() and add a test for this method.

kenjis commented:
We tried to match the PDO method signatures in the PHP manual as much as possible.

kenjis commented:
@srjlewis Thank you!

kenjis commented:
Close #192
